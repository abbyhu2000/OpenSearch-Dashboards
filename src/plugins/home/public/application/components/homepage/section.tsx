/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { FC, useState, useMemo } from 'react';
import { i18n } from '@osd/i18n';
import { EuiPanel, EuiButtonIcon, EuiTitle, EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
import { RenderFn, Section as SectionType } from '../../../services/section_type/section_type';
import { LazyRender } from './lazy_render';

interface Props {
  render: RenderFn;
  title: SectionType['title'];
  headerComponent?: React.ReactNode;
}

export const Section: FC<Props> = ({ render, title, headerComponent }) => {
  const [isExpanded, setExpanded] = useState(true);
  const toggleExpanded = () => setExpanded((expanded) => !expanded);

  const memoizedContent = useMemo(
    () => (
      <EuiFlexGroup direction="row" data-test-subj="homepageSectionContent">
        <EuiFlexItem grow={3}>
          <LazyRender render={render} />
        </EuiFlexItem>
      </EuiFlexGroup>
    ),
    [render]
  );

  return (
    <EuiPanel
      hasBorder={false}
      hasShadow={false}
      color="transparent"
      data-test-subj="homepageSection"
    >
      <EuiFlexGroup direction="row" alignItems="center" gutterSize="s" responsive={false}>
        <EuiFlexItem grow>
          <EuiTitle size="m">
            <h2>{title}</h2>
          </EuiTitle>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>{headerComponent}</EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButtonIcon
            iconType={isExpanded ? 'arrowDown' : 'arrowRight'}
            onClick={toggleExpanded}
            size="s"
            iconSize="m"
            color="text"
            aria-label={
              isExpanded
                ? i18n.translate('home.section.collapse', { defaultMessage: 'Collapse section' })
                : i18n.translate('home.section.expand', { defaultMessage: 'Expand section' })
            }
          />
        </EuiFlexItem>
      </EuiFlexGroup>
      {isExpanded && memoizedContent}
    </EuiPanel>
  );
};
