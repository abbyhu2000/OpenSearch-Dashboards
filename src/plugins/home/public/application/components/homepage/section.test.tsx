/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiButtonIcon, EuiLink } from '@elastic/eui';
import { mount } from 'enzyme';
import { Section } from './section';

let title: string;
let headerComponent: React.ReactNode;
let render: any;
let component: any;

describe('Home page sections', () => {
  beforeAll(() => {
    title = 'Section title';
    headerComponent = (
      <EuiLink href="https" target="_blank" className="section-links">
        See all documentation
      </EuiLink>
    );
    render = jest.fn();
    component = mount(<Section title={title} render={render} headerComponent={headerComponent} />);
  });

  it('is rendered with title and header component', () => {
    expect(component).toMatchSnapshot();
    expect(component.find(EuiButtonIcon).prop('iconType')).toBe('arrowDown');
  });

  it('can be collapse on and off', () => {
    component.find(EuiButtonIcon).simulate('click');
    expect(component.find(EuiButtonIcon).prop('iconType')).toBe('arrowRight');
  });
});
