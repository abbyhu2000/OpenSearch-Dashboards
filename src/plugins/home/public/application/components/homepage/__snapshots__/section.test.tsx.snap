// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page sections is rendered with title and header component 1`] = `
<Section
  headerComponent={
    <EuiLink
      className="section-links"
      href="https"
      target="_blank"
    >
      See all documentation
    </EuiLink>
  }
  render={
    [MockFunction] {
      "calls": Array [
        Array [
          <div
            style="display: contents;"
          />,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  title="Section title"
>
  <EuiPanel
    color="transparent"
    data-test-subj="homepageSection"
    hasBorder={false}
    hasShadow={false}
  >
    <div
      className="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--transparent euiPanel--noShadow euiPanel--noBorder"
      data-test-subj="homepageSection"
    >
      <EuiFlexGroup
        alignItems="center"
        direction="row"
        gutterSize="s"
        responsive={false}
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
        >
          <EuiFlexItem
            grow={true}
          >
            <div
              className="euiFlexItem"
            >
              <EuiTitle
                size="m"
              >
                <h2
                  className="euiTitle euiTitle--medium"
                >
                  Section title
                </h2>
              </EuiTitle>
            </div>
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
          >
            <div
              className="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <EuiLink
                className="section-links"
                href="https"
                target="_blank"
              >
                <a
                  className="euiLink euiLink--primary section-links"
                  href="https"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  See all documentation
                  <EuiIcon
                    aria-label="External link"
                    className="euiLink__externalIcon"
                    size="s"
                    type="popout"
                  >
                    <span
                      aria-label="External link"
                      className="euiLink__externalIcon"
                      data-euiicon-type="popout"
                      size="s"
                    />
                  </EuiIcon>
                  <EuiScreenReaderOnly>
                    <span
                      className="euiScreenReaderOnly"
                    >
                      <EuiI18n
                        default="(opens in a new tab or window)"
                        token="euiLink.newTarget.screenReaderOnlyText"
                      >
                        (opens in a new tab or window)
                      </EuiI18n>
                    </span>
                  </EuiScreenReaderOnly>
                </a>
              </EuiLink>
            </div>
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
          >
            <div
              className="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <EuiButtonIcon
                aria-label="Collapse section"
                color="text"
                iconSize="m"
                iconType="arrowDown"
                onClick={[Function]}
                size="s"
              >
                <button
                  aria-label="Collapse section"
                  className="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--small"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <EuiIcon
                    aria-hidden="true"
                    className="euiButtonIcon__icon"
                    color="inherit"
                    size="m"
                    type="arrowDown"
                  >
                    <span
                      aria-hidden="true"
                      className="euiButtonIcon__icon"
                      color="inherit"
                      data-euiicon-type="arrowDown"
                      size="m"
                    />
                  </EuiIcon>
                </button>
              </EuiButtonIcon>
            </div>
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
      <EuiFlexGroup
        data-test-subj="homepageSectionContent"
        direction="row"
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          data-test-subj="homepageSectionContent"
        >
          <EuiFlexItem
            grow={3}
          >
            <div
              className="euiFlexItem euiFlexItem--flexGrow3"
            >
              <LazyRender
                render={
                  [MockFunction] {
                    "calls": Array [
                      Array [
                        <div
                          style="display: contents;"
                        />,
                      ],
                    ],
                    "results": Array [
                      Object {
                        "type": "return",
                        "value": undefined,
                      },
                    ],
                  }
                }
              >
                <div
                  style={
                    Object {
                      "display": "contents",
                    }
                  }
                />
              </LazyRender>
            </div>
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
    </div>
  </EuiPanel>
</Section>
`;
